caffe2/serialize/intermediate_model.h

```diff
@@ -0,0 +1,318 @@
+  std::string nodeName;
+  std::vector<std::string> extraInfo;
+};
+
+class IntermediateTensor final {
+ public:
+  // constructor
+  IntermediateTensor() = default;
+
+  // update functions
+  void update(caffe2::TensorProto* tensor_proto,
+      std::unordered_map<int64_t, at::DataPtr>* id_tensor) {
+    CAFFE_ENFORCE(tensor_proto->has_data_type(), "no data_type in TensorProto!");
+    dataType_ = tensor_proto->data_type();
+    for (int i = 0; i < tensor_proto->dims_size()) {
+      dims_.push_back(tensor_proto->dims(i));
+    }
+    CAFFE_ENFORCE(tensor_proto->has_name(), "no name in TensorProto!");
+    name_ = tensor_proto->name();
+    if (tensor_proto->has_device_detail()) {
+      const auto& device_detail = tensor_proto->device_detail();
+      deviceDetail_.deviceType = device_detail.device_type;
+      if (device_detail.has_device_id()) {
+        deviceDetail_.hasDeviceId = true;
+        deviceDetail_.devicedId = device_detail.device_id();
+      }
+      if (device_detail.has_random_seed()) {
+        deviceDetail_.hasRandomSeed = true;
+        deviceDetail_.randomSeed = device_detail.random_seed();
+      }
+      if (device_detail.has_node_name()) {
+        deviceDetail_.hasNodeName = true;
+        deviceDetail_.nodeName = device_detail.nodeName();
+      }
+      for (int i = 0; i < device_detail.extra_info_size(); ++i) {
+        deviceDetail_.extraInfo.emplace_back(device_detail.extra_info(i));
+      }
+    }
+    CAFFE_ENFORCE(tensor_proto->has_storage_type(), "no storage_type in TensorProto!");
+    int64_t storage_type = tensor_proto->storage_type();
+    switch (storage_type) {
+      case TensorProto_StorageType_TYPED:
+        // TODO
+        CAFFE_THROW("Not implemented!");
+      case TensorProto_StorageType_RAW:
+        // TODO create the data type
+        CAFFE_THROW("Not implemented!");
+      case TensorProto_StorageType_EXTERNAL:
+        CAFFE_ENFORCE(tensor_proto->has_external_data(), "storage type is EXTERNAL, "
+            "but no external_data in TensorProto!");
+        const auto& external_data = tensor_proto->external_data();
+        if (external_data->has_offset()) {
+          offset_ = external_data->offset();
+        }
+        for (int i = 0; i < external_data.strides_size(); ++i) {
+          strides_->push_back(external_data.strides(i));
+        }
+        CAFFE_ENFORCE(external_data->has_source_type(), "no source_type in TensorProto!");
+        int64_t source_type = external_data->source_type();
+        if (source_type == ExternalDataProto_SourceType::ExternalDataProto_SourceType_INLINE_CONTAINER) {
+          CAFFE_ENFORCE(external_data->has_record_id(), "no record_id in ExternalDataProto!");
+          int64_t record_id = external_data->record_id();
+          auto it = id_tensor.find(record_id);
+          if (it == id_tensor.end()) {
+            CAFFE_THROW("Tensor's data is missing, tensor name is ", name_);
+          }
+          dataPtr_ = std::move(it->second);
```

Comment:
how does it work with shared storages? you'd move out the storage first time (not sure whether it was tested or not in the prototype, but we should be able to serialize aliasing of storages when several tensors point to the same storage and we serialize storage once)